pkg/selinux: various fixes

iaguis opened the pull request:
TestSELinux was not setting the MCS dir which caused the paths of every
MCS to be in the root of the file system (the path is calculated as
`fmt.Sprintf("%s/%s", mcsdir, mcs)` and `mcsdir` was the empty string by
default). This was making the test to fail when it runs as non-root.

The test failure was non-obvious because of the way the algorithm to get
a unique MCS is implemented. We first try to create the MCS file with a
random category pair and, if it doesn't work, we continue with a
different random category pair. Since we can't write to the MCS
directory, we'll continue doing this until the test times out and we
don't print any errors when we can't create the MCS file.

Let's clean this package a bit.

iaguis commented:
Incorporated #2724 <3

jonboulle commented:
LGTM but could do with another review, @lucab ?

dgonyeo commented on pkg/selinux/selinux.go:
```diff
@@ -361,19 +361,19 @@ func FreeLxcContexts(scon string) {
 	in, err := os.Open(lxcPath)
 	if err != nil {
-		return "", ""
+		return "", "", nil
```
If there's an error, return nil?

iaguis replied:
Don't ask me! Haha

I guess if there's no lxc_contexts then we just return empty. Callers seem to handle this by doing nothing.

dgonyeo commented on pkg/selinux/selinux_test.go:
```diff
@@ -18,66 +19,80 @@
+		panic(fmt.Sprintf("unable to create temp dir: %v", err))
+	}
+	defer os.RemoveAll(dir)
+	_ = selinux.SetMCSDir(dir)
```
Why not check the error?

iaguis replied:
We just created the directory and the only error that `SetMCSDir()` calls `MkdirAll()`, I can check it anyway, I guess it doesn't hurt.

dgonyeo commented:
Aside from two cases where the error checking looks weird, LGTM.

lucab commented on pkg/selinux/selinux.go:
```diff
@@ -409,19 +409,21 @@ func GetLxcContexts() (processLabel string, fileLabel string) {
 
 	if processLabel == "" || fileLabel == "" {
-		return "", ""
+		return "", "", nil
```
Isn't this one (and the two above) an error condition? Why are they returning an nil error?

iaguis replied:
I assumed that in those cases (no SELinux, no lxc_contexts file and invalid content on lxc_contexts) `GetLxcContexts()` is a no-op

iaguis replied:
@mjg59?

mjg59 replied:
Yeah I think that's reasonable. It's valid for selinux to be enabled and for us to have no lxc_contexts - there's no standard for selinux policy. In that case we should just continue without it.

jonboulle replied:
Luca raised the reasonable point put of band that we might want to move to
using the actual latest selinux package as a dependency so we don't have to
maintain our own (assuming of course that it's in a more sane state by now
- perhaps it isn't)

Matthew Garrett <notifications@github.com> schrieb am Di., 31. Mai 2016,
19:04:

> In pkg/selinux/selinux.go
> <https://github.com/coreos/rkt/pull/2723#discussion_r65222667>:
>
> > @@ -409,19 +409,21 @@ func GetLxcContexts() (processLabel string, fileLabel string) {
> >  	}
> >
> >  	if processLabel == "" || fileLabel == "" {
> > -		return "", ""
> > +		return "", "", nil
>
> Yeah I think that's reasonable. It's valid for selinux to be enabled and
> for us to have no lxc_contexts - there's no standard for selinux policy. In
> that case we should just continue without it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/coreos/rkt/pull/2723/files/438c1e7197aedb6d83b10beff5cd6b8947a7c20a#r65222667>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACewNz07_v2Nar1pfbHgzynOsM8iL0aIks5qHGofgaJpZM4IqV-A>
> .
>

iaguis replied:
Hmmm... https://github.com/opencontainers/runc/blob/v0.1.1/libcontainer/selinux/selinux.go#L307

lucab commented:
I would not merge as is. A bunch of remarks:
 * error reporting in general looks weird
 * this is effectively switching from crypto/rand to math/rand. Commit should mention somewhere why it makes sense given current usage
 * this looks like we are further forking an already smelly code, which is not rkt specific. Can't this be externalized to something better instead?

mjg59 commented:
The selinux package is part of libcontainer, so we can either pull in all of libcontainer or live with a fork.

iaguis commented:
I addressed https://github.com/coreos/rkt/pull/2723#discussion_r65220880 and reverted to using `crypto/rand`.

I opened https://github.com/coreos/rkt/issues/2727 to discuss externalizing pkg/selinux.

lucab commented:
This looks already better to me. Let's squash this changes in two logical commits (test fixes and code package cleanup) and I'll have another pass at it. That way it should also be easier to move upstream and ignite the discussion there.

iaguis commented:
Done.

iaguis commented:
Fixes https://github.com/coreos/rkt/issues/2711

lucab commented:
LGTM now. I'm just bouncing this back to @jonboulle for a final double check given the multiple reworks.

jonboulle commented:
LGTM
